include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,374 @@
+  void set_error(E&& err) noexcept {
+    unifex::set_error(
+        std::move(*genCoro_.promise().receiverOpt_), std::forward<E>(err));
+  }
+};
+
+// Note this is also used in final_suspend, since it handles moving
+// back to the consumer's scheduler, instead of using at_coroutine_exit
+// as in unifex::task<>
+template <typename T>
+class async_generator_yield_operation {
+public:
+  using value_type = std::remove_reference_t<T>;
+
+  async_generator_yield_operation(std::optional<value_type> value = {}) noexcept
+    : value_{std::move(value)} {}
+
+  bool await_ready() const noexcept { return false; }
+
+  template <typename Promise>
+  void await_suspend([[maybe_unused]] unifex::coro::coroutine_handle<Promise>
+                         genCoro) noexcept {
+    const auto& consumerSched = genCoro.promise().consumerSched_;
+    if (unifex::get_scheduler(genCoro.promise()) != consumerSched) {
+      genCoro.promise().rescheduleOpSt_ = unifex::connect(
+          unifex::schedule(consumerSched),
+          reschedule_receiver<Promise>{std::move(value_), genCoro});
+      unifex::start(*genCoro.promise().rescheduleOpSt_);
+      return;
+    }
+
+    if (value_) {
+      unifex::set_value(std::move(*genCoro.promise().receiverOpt_), *value_);
+    } else {
+      unifex::set_done(std::move(*genCoro.promise().receiverOpt_));
+    }
+  }
+
+  void await_resume() noexcept {}
+
+private:
+  std::optional<value_type> value_;
+};
+
+template <typename T>
+class async_generator_promise {
+public:
+  using value_type = std::remove_reference_t<T>;
+  using Promise = async_generator_promise<T>;
+
+  async_generator_promise() noexcept : exception_(nullptr) {
+    // Other variables left intentionally uninitialised as they're
+    // only referenced in certain states by which time they should
+    // have been initialised.
+  }
+
+  async_generator_promise(const async_generator_promise& other) = delete;
+  async_generator_promise&
+  operator=(const async_generator_promise& other) = delete;
+
+  std::suspend_always initial_suspend() const noexcept { return {}; }
+
+  void unhandled_exception() noexcept { exception_ = std::current_exception(); }
+
+  void return_void() noexcept {}
+
+  void rethrow_if_unhandled_exception() {
+    if (exception_) {
+      std::rethrow_exception(std::move(exception_));
+    }
+  }
+
+  // This is the CP needed for other things in the ecosystem to be able to
+  // "access" the current scheduler of our promise.
+  friend unifex::any_scheduler tag_invoke(
+      unifex::tag_t<unifex::get_scheduler>,
+      const async_generator_promise& p) noexcept {
+    return *p.sched_;
+  }
+
+  // This is needed for at_coroutine_exit to do the async clean up
+  friend unifex::continuation_handle<> tag_invoke(
+      const unifex::tag_t<unifex::exchange_continuation>&,
+      async_generator_promise& p,
+      unifex::continuation_handle<> action) noexcept {
+    return std::exchange(p.continuation_, std::move(action));
+  }
```

Comment:
I think this is the wrong implementation for `at_coroutine_exit`. Instead, the generator object should have an initially-null `coroutine_handle<>` that points to the top of a stack of continuations to be awaited in `cleanup()`. Invoking `at_coroutine_exit` inside a generator should push the new coroutine onto the stack. Then, `cleanup()` should run the stack of continuations (if it's non-empty) before completing with done.

Alternatively, perhaps the handle shouldn't be initially-null, but the result of invoking this coroutine:
```
nothrow_task<void> cleanup() {
  co_await just_done();
}
```
then you always have a non-empty stack.

Reply:
@ispeters this is actually a leftover from when I was using `at_coroutine_exit`. There were a few things, for which I needed it:

1) awaiting `join()` on the async scope
2) rescheduling back onto the consumer's scheduler (basically the equivalent of what is done in `unifex::task<>` on coroutine exit.

Right now 1) is no longer needed.

2) I believe is handled within the `async_generator_yield_operation<>`. This one has the scheduler affinity logic within its `await_suspend` - if we aren't on the correct scheduler, we kick off a `unifex::reschedule` to go back to it.

Finally, we return this awaiter both in `yield_value()` and `final_suspend`. So this should automatically handle the rescheduling during the generator's coroutine exit. 

Let me know what you think, but it seems we don't need `at_coroutine_exit` at all here.

Reply:
I see what you mean and also your very original comment explains the same thing (some stuff starts making sense after a while with this code base... 😄).

Reply:
I'm not 100% certain this is going to work, but here's how I _think_ it'll work.

[`at_coroutine_exit`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/at_coroutine_exit.hpp#L317-L320) returns a [`_cleanup_task<>`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/at_coroutine_exit.hpp#L266). `_cleanup_task<>`'s implementation of [`await_suspend`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/at_coroutine_exit.hpp#L279-L290) [invokes `exchange_continuation`](https://fburl.com/jwpk5nds) on the calling coroutine's promise. So, I _think_ you can implement `exchange_continuation` in such a way that each call pushes another clean-up task onto the stack of things to resume from `cleanup()`.